stackslib/src/chainstate/burn/atc.rs

```diff
@@ -0,0 +1,1403 @@
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use stacks_common::util::uint::Uint256;
+
+use crate::stacks_common::util::uint::BitArray;
+
+/// A fixed-point numerical representation for ATC.  The integer and fractional parts are both 64
+/// bits.  Internally, this is a Uint256 so that safe addition and multiplication can be done.
+///
+/// Bits 0-63 are the fraction.
+/// Bits 64-127 are the integer.
+/// Bits 128-256 are 0's to facilitate safe addition and multiplication.
+///
+/// The reasons we use this instead of f64 for ATC calculations are as follows:
+/// * This avoids unrepresentable states, like NaN or +/- INF
+/// * This avoids ambiguous states, like +0.0 and -0.0.
+/// * This integrates better into the sortition-sampling system, which uses a u256 to represent a
+/// probability range (which is what this is going to be used for)
+#[derive(Debug, Clone, PartialEq, Copy, Eq, Hash)]
+pub(crate) struct AtcRational(pub(crate) Uint256);
+impl AtcRational {
+    /// Construct from a fraction (numerator and denominator)
+    pub fn frac(num: u64, den: u64) -> Self {
+        Self((Uint256::from_u64(num) << 64) / Uint256::from_u64(den))
+    }
+
+    /// 0 value
+    pub fn zero() -> Self {
+        Self(Uint256::zero())
+    }
+
+    /// 1 value
+    pub fn one() -> Self {
+        Self(Uint256::one() << 64)
+    }
+
+    /// largest value less than 1
+    pub fn one_sup() -> Self {
+        Self((Uint256::one() << 64) - Uint256::from_u64(1))
+    }
+
+    /// Largest possible value (corresponds to u64::MAX.u64::MAX)
+    pub fn max() -> Self {
+        Self((Uint256::from_u64(u64::MAX) << 64) | Uint256::from_u64(u64::MAX))
+    }
+
+    /// Get integer part
+    pub fn ipart(&self) -> u64 {
+        (self.0 >> 64).low_u64()
+    }
+
+    /// Checked addition
+    pub fn add(&self, other: &AtcRational) -> Option<Self> {
+        // NOTE: this is always safe since u128::MAX + u128::MAX < Uint256::max()
+        let sum = AtcRational(self.0 + other.0);
+        if sum.0 > Self::max().0 {
+            return None;
+        }
+        Some(sum)
+    }
+
+    /// Checked subtraction
+    pub fn sub(&self, other: &AtcRational) -> Option<Self> {
+        if self.0 < other.0 {
+            return None;
+        }
+        Some(AtcRational(self.0 - other.0))
+    }
+
+    /// Checked multiplication
+    pub fn mul(&self, other: &AtcRational) -> Option<Self> {
+        // NOTE: this is always safe since u128::MAX * u128::MAX < Uint256::max()
+        let prod = AtcRational((self.0 * other.0) >> 64);
+        if prod.0 > Self::max().0 {
```

Comment:
Can you move this check to a function like `is_overflowed(&self)`?

Reply:
Sure